From 8495f3ff6f49a1d923ee24d0d0e625783220ba6d Mon Sep 17 00:00:00 2001
From: Michael Simons <msimons@microsoft.com>
Date: Mon, 26 Jul 2021 16:33:22 +0000
Subject: [PATCH] Fix ambiguous DocumentBasedFixAllProvider reference

This addresses ambiguous reference issues related to DocumentBasedFixAllProvider.

Example error:
src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/Fixers/PreferIsKindFix.cs(51,53): error CS0104: 'DocumentBasedFixAllProvider' is an ambiguous reference between 'Analyzer.Utilities.DocumentBasedFixAllProvider' and 'Microsoft.CodeAnalysis.CodeFixes.DocumentBasedFixAllProvider'

This is a result of https://github.com/dotnet/roslyn/commit/98e8cb016993a9bf3896e24f7d02947f6046ef97#diff-dd947268a990618d595914e376b02f059f6c761f61ce77eaa2bbf270f41ebbfc which publically exposed DocumentBasedFixAllProvider in Roslyn.

'Microsoft.CodeAnalysis.CodeFixes.DocumentBasedFixAllProvider' is not compatible with `Analyzer.Utilities.DocumentBasedFixAllProvider`, therefore the patch is to fully qualify the type reference to the roslyn-analyzer's type.
---
 .../Core/MetaAnalyzers/Fixers/PreferIsKindFix.cs                | 2 +-
 .../ApiDesignGuidelines/DoNotDirectlyAwaitATask.Fixer.cs        | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/Fixers/PreferIsKindFix.cs b/src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/Fixers/PreferIsKindFix.cs
index f344b92fd..bc815d973 100644
--- a/src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/Fixers/PreferIsKindFix.cs
+++ b/src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/Fixers/PreferIsKindFix.cs
@@ -48,7 +48,7 @@ private async Task<Document> ConvertKindToIsKindAsync(Document document, TextSpa
 
         protected abstract void FixDiagnostic(DocumentEditor editor, SyntaxNode nodeToFix);
 
-        private sealed class CustomFixAllProvider : DocumentBasedFixAllProvider
+        private sealed class CustomFixAllProvider : Analyzer.Utilities.DocumentBasedFixAllProvider
         {
             private readonly PreferIsKindFix _fixer;
 
diff --git a/src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/DoNotDirectlyAwaitATask.Fixer.cs b/src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/DoNotDirectlyAwaitATask.Fixer.cs
index a6826eaba..ba0b576dd 100644
--- a/src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/DoNotDirectlyAwaitATask.Fixer.cs
+++ b/src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/DoNotDirectlyAwaitATask.Fixer.cs
@@ -80,7 +80,7 @@ private class MyCodeAction : DocumentChangeAction
             }
         }
 
-        private sealed class CustomFixAllProvider : DocumentBasedFixAllProvider
+        private sealed class CustomFixAllProvider : Analyzer.Utilities.DocumentBasedFixAllProvider
         {
             public static readonly CustomFixAllProvider Instance = new();
 
-- 
2.29.2

